File: src/sys/dotnet/fan/sys/Err.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   24 Sep 06  Andy Frank  Creation
//

using System;
using System.Diagnostics;
using System.IO;
using System.Net.Sockets;
using System.Reflection;
using System.Text;
using System.Threading;

namespace Fan.Sys
{
  /// <summary>
  /// Err.
  /// </summary>
  public class Err : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // C# to Fantom Mapping
  //////////////////////////////////////////////////////////////////////////

    /// <summary>
    /// Map a java exception to it's Fantom Err counter part.  Common runtime
    /// exceptions are mapped into explicit Fantom types.  Otherwise we just
    /// wrap the exception with a generic Err.
    /// </summary>
    public static Err make(Exception ex)
    {
      if (ex == null) return null;
      if (ex is Val)                         return ((Val)ex).m_err;
      if (ex is NullReferenceException)      return new NullErr(ex);
      if (ex is InvalidCastException)        return new CastErr(ex);
      if (ex is IndexOutOfRangeException)    return new IndexErr(ex);
      if (ex is ArgumentOutOfRangeException) return new IndexErr(ex);
      if (ex is ArgumentException)           return new ArgErr(ex);
      if (ex is IOException)                 return new IOErr(ex);
      if (ex is SocketException)             return new IOErr(ex); // TODO?
      if (ex is ThreadInterruptedException)  return new InterruptedErr(ex);
      if (ex is NotSupportedException)       return new UnsupportedErr(ex);
      return new Err(new Err.Val(), ex);
    }

    /// <summary>
    /// This method is used by FCodeEmit to generate extra entries in the
    /// exception table - for example if fcode says to trap NullErr, then
    /// we also need to trap System.NullReferenceException.  Basically this
    /// is the inverse of the mapping done in make(Exception).
    /// </summary>
    public static string fanToDotnet(string ftype)
    {
      // TODO - need to trap as well...
      //if (ftype == "Fan.Sys.IndexErr")) return "ArgumentOutOfRangeException";

      if (ftype == "Fan.Sys.NullErr")  return "System.NullReferenceException";
      if (ftype == "Fan.Sys.CastErr")  return "System.InvalidCastException";
      if (ftype == "Fan.Sys.IndexErr") return "System.IndexOutOfRangeException";
      if (ftype == "Fan.Sys.ArgErr")   return "System.ArgumentException";
      if (ftype == "Fan.Sys.IOErr")    return "System.IO.IOException";
      if (ftype == "Fan.Sys.InterruptedErr") return "System.Threading.ThreadInterruptedException";
      if (ftype == "Fan.Sys.UnsupportedErr") return "System.NotSupportedException";
      return null;
    }

  //////////////////////////////////////////////////////////////////////////
  // C# Convenience
  //////////////////////////////////////////////////////////////////////////

    public static Err make(string msg, Exception e) { return make(msg, make(e)); }

  //////////////////////////////////////////////////////////////////////////
  // Fantom Constructors
  //////////////////////////////////////////////////////////////////////////

    public static Err make() { return make("", (Err)null); }
    public static Err make(string msg) { return make(msg, (Err)null); }
    public static Err make(string msg, Err cause)
    {
      Err err = new Err(new Err.Val());
      make_(err, msg, cause);
      return err;
    }

    public static void make_(Err self) { make_(self, null);  }
    public static void make_(Err self, string msg) { make_(self, msg, null); }
    public static void make_(Err self, string msg, Err cause)
    {
      if (msg == null) throw NullErr.make("msg is null").val;
      self.m_msg = msg;
      self.m_cause = cause;
    }

  //////////////////////////////////////////////////////////////////////////
  // C# Constructors
  //////////////////////////////////////////////////////////////////////////

    public Err() {}

    /// <summary>
    /// All subclasses must call this constructor with their
    /// typed Val exception which the real .NET exception we use.
    /// </summary>
    public Err(Val val)
    {
      this.val = val;
      val.m_err = this;
    }

    /// <summary>
    /// This constructor is used by special subclasses which provide
    /// a transparent mapping between .NET and Fantom exception types.
    /// </summary>
    public Err(Val val, Exception actual)
    {
      this.val = val;
      val.m_err = this;
      this.m_actual = actual;
      this.m_msg = actual.Message;
    }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public string msg()
    {
      return m_msg;
    }

    public Err cause()
    {
      return m_cause;
    }

    public Err trace() { return trace(Env.cur().@out(), null, 0, true); }
    public Err trace(OutStream @out) { return trace(@out, null, 0, true); }
    public Err trace(OutStream @out, Map opt) { return trace(@out, opt, 0, true); }
    public Err trace(OutStream @out, Map opt, int indent, bool useActual)
    {
      Exception ex = m_actual != null && useActual ? m_actual : val;
      dumpStack(toStr(), ex, @out, indent);
      if (m_cause != null)
      {
        @out.printLine("Cause:");
        m_cause.trace(@out, opt, indent+2, useActual);
      }
      return this;
    }

    public string traceToStr()
    {
      Buf buf = new MemBuf(1024);
      trace(buf.@out());
      return buf.flip().readAllStr();
    }

    public override Type @typeof()
    {
      return Sys.ErrType;
    }

    public override string toStr()
    {
      if (m_msg == null || m_msg.Length == 0)
        return @typeof().qname();
      else
        return @typeof().qname() + ": " + m_msg;
    }

  //////////////////////////////////////////////////////////////////////////
  // Rebasing
  //////////////////////////////////////////////////////////////////////////

    public Val rebase()
    {
      m_actual = new RebaseException();
      return val;
    }

    public class RebaseException : Exception
    {
    }

  //////////////////////////////////////////////////////////////////////////
  // Val
  //////////////////////////////////////////////////////////////////////////

    public class Val : Exception
    {
      public override string Message { get { return m_err.ToString(); }}
      public Err err() { return m_err; }
      public Err m_err;
    }

  //////////////////////////////////////////////////////////////////////////
  // Tracing
  //////////////////////////////////////////////////////////////////////////

    /// <summary>
    /// Dump a readable stack trace.
    /// </summary>
    public static void dumpStack(Exception err) { dumpStack(err.Message, err, 0); }
    public static void dumpStack(string msg, Exception err, int indent)
    {
      StringWriter w = new StringWriter();
      doDumpStack(msg, err, indent, w);
      System.Console.Write(w.ToString());
    }

    /// <summary>
    /// Dump a readable stack trace.
    /// </summary>
    public static void dumpStack(Exception err, OutStream @out) { dumpStack(err.Message, err, @out, 0); }
    public static void dumpStack(string msg, Exception err, OutStream @out, int indent)
    {
      StringWriter w = new StringWriter();
      doDumpStack(msg, err, indent, w);
      @out.writeChars(w.ToString()).flush();
    }

    static void doDumpStack(string msg, Exception err, int depth, StringWriter w)
    {
      // message
      for (int sp=0; sp<depth; sp++) w.Write(" ");
      if (!(err is Err.Val) && msg == err.Message) w.Write(err.GetType() + ": ");
      w.WriteLine(msg);

      // stack
      string stack = err.StackTrace;
      if (err is Err.Val)
      {
        Err e = ((Err.Val)err).err();
        if (e.m_stack != null) stack = e.m_stack;
      }
      if (stack != null)
      {
        string[] lines = stack.Split('\n');
        for (int i=0; i<lines.Length; i++)
        {
          // TODO - could be *way* more efficient

          string s = lines[i].Trim();
          int parOpen  = s.IndexOf('(');
          int parClose = s.IndexOf(')', parOpen);

          string source = s.Substring(parClose+1, s.Length-parClose-1);
          if (source == "") source = "Unknown Source";
          else
          {
            source = source.Substring(4);
            int index = source.LastIndexOf("\\");
            if (index != -1) source = source.Substring(index+1);
            index = source.LastIndexOf(":line");
            source = source.Substring(0, index+1) + source.Substring(index+6);
          }

          string target = s.Substring(0, parOpen);
          if (target.StartsWith("at Fan."))
          {
            int a = target.IndexOf(".", 7);
            int b = target.IndexOf(".", a+1);
            string pod  = target.Substring(7, a-7);
            string type = target.Substring(a+1, b-a-1);
            string meth = target.Substring(b+1);

            // check for closures
            int dollar1 = type.IndexOf('$');
            int dollar2 = dollar1 < 0 ? -1 : type.IndexOf('$', dollar1+1);
            if (dollar2 > 0)
            {
              // don't print callX for closures
              if (meth.StartsWith("call")) continue;
              // remap closure class back to original method
              if (meth.StartsWith("doCall"))
              {
                meth = type.Substring(dollar1+1, dollar2-dollar1-1);
                type = type.Substring(0, dollar1);
              }
            }

            target = FanStr.decapitalize(pod) + "::" + type + "." + meth;
          }

          for (int sp=0; sp<depth; sp++) w.Write(" ");
          w.Write("  ");
          w.Write(target);
          w.Write(" (");
          w.Write(source);
          w.Write(")");
          w.Write("\n");
        }
      }
      // inner exception
      Exception cause = err.InnerException;
      if (cause != null) doDumpStack(msg, cause, depth+1, w);
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    public readonly Val val;
    internal string m_msg;
    internal Err m_cause = null;
    internal Exception m_actual;
    internal string m_stack;       // only used for Method.invoke()

  }
}

File: src/sys/java/fan/sys/Err.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   4 Jan 06  Brian Frank  Creation
//
package fan.sys;

import java.io.*;
import java.util.*;

/**
 * Err
 */
public class Err
  extends RuntimeException
{

//////////////////////////////////////////////////////////////////////////
// Java to Fantom Mapping
//////////////////////////////////////////////////////////////////////////

  /**
   * Map a java exception to it's Fantom Err counter part.  Common runtime
   * exceptions are mapped into explicit Fantom types.  Otherwise we just
   * wrap the exception with a generic Err.
   */
  public static Err make(Throwable ex)
  {
    // NOTE: everything in this list must be synchronized
    // with the mapping used below for FCodeEmit.errTable()
    // and tested in TryTest
    if (ex == null) return null;
    if (ex instanceof Err) return (Err)ex;
    if (ex instanceof NullPointerException)      return new NullErr(ex);
    if (ex instanceof ClassCastException)        return new CastErr(ex);
    if (ex instanceof IndexOutOfBoundsException) return new IndexErr(ex);
    if (ex instanceof IllegalArgumentException)  return new ArgErr(ex);
    if (ex instanceof IOException)               return new IOErr(ex);
    if (ex instanceof InterruptedException)      return new InterruptedErr(ex);
    if (ex instanceof UnsupportedOperationException)  return new UnsupportedErr(ex);
    return new Err(ex);
  }

  /**
   * This method is used by FCodeEmit to generate extra entries in the
   * exception table - for example if fcode says to trap NullErr, then
   * we also need to trap java.lang.NullPointerException.  Basically this
   * is the inverse of the mapping done in make(Throwable).
   */
  public static String fanToJava(String jtype)
  {
    if (jtype.equals("fan/sys/NullErr"))  return "java/lang/NullPointerException";
    if (jtype.equals("fan/sys/CastErr"))  return "java/lang/ClassCastException";
    if (jtype.equals("fan/sys/IndexErr")) return "java/lang/IndexOutOfBoundsException";
    if (jtype.equals("fan/sys/ArgErr"))   return "java/lang/IllegalArgumentException";
    if (jtype.equals("fan/sys/IOErr"))    return "java/io/IOException";
    if (jtype.equals("fan/sys/InterruptedErr")) return "java/lang/InterruptedException";
    if (jtype.equals("fan/sys/UnsupportedErr")) return "java/lang/UnsupportedOperationException";
    return null;
  }

//////////////////////////////////////////////////////////////////////////
// Java Convenience
//////////////////////////////////////////////////////////////////////////

  public static Err make(String msg, Throwable e) { return make(msg, make(e)); }

//////////////////////////////////////////////////////////////////////////
// Fantom Constructors
//////////////////////////////////////////////////////////////////////////

  public static Err make() { return make("", (Err)null); }
  public static Err make(String msg) { return make(msg, (Err)null); }
  public static Err make(String msg, Err cause)
  {
    Err err = new Err();
    make$(err, msg, cause);
    return err;
  }

  public static void make$(Err self) { make$(self, null);  }
  public static void make$(Err self, String msg) { make$(self, msg, null); }
  public static void make$(Err self, String msg, Err cause)
  {
    if (msg == null) throw NullErr.make("msg is null");
    self.msg = msg;
    self.cause = cause;
  }

//////////////////////////////////////////////////////////////////////////
// Java Constructors
//////////////////////////////////////////////////////////////////////////

  /**
   * This constructor is used by special subclasses which provide
   * a transparent mapping between Java and Fantom exception types.
   */
  public Err(Throwable actual)
  {
    this.actual = actual;
    this.msg = actual.toString();
  }

  /**
   * No argument constructor.
   */
  public Err()
  {
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public String msg() { return msg; }

  public Err cause() { return cause; }

  public Type typeof() { return Sys.ErrType; }

  public String toStr()
  {
    // wrap with try block to safely handle boot-strap problems
    String qname;
    try { qname = typeof().qname(); }
    catch (Throwable e) { qname = getClass().getName(); }

    if (msg == null || msg.length() == 0)
      return qname;
    else
      return qname + ": " + msg;
  }

//////////////////////////////////////////////////////////////////////////
// Trace
//////////////////////////////////////////////////////////////////////////

  public Err trace() { return trace(Env.cur().out(), null, 0, true); }
  public Err trace(OutStream out) { return trace(out, null, 0, true); }
  public Err trace(OutStream out, Map opt) { return trace(out, opt, 0, true); }
  public Err trace(OutStream out, Map opt, int indent, boolean useActual)
  {
    // map exception to stack trace
    Throwable ex = actual != null && useActual ? actual : this;
    StackTraceElement[] elems = ex.getStackTrace();

    // extract options
    int maxDepth = 20;
    if (opt != null)
    {
      Long m = (Long)opt.get("maxDepth");
      if (m != null) { maxDepth = m.longValue() > Integer.MAX_VALUE ? Integer.MAX_VALUE : m.intValue(); }
    }

    // skip calls to make the Err itself
    int start = 0;
    for (; start<elems.length; ++start)
    {
      StackTraceElement elem = elems[start];
      if (elem.getClassName().endsWith("Err"))
      {
        String m = elem.getMethodName();
        if (m.equals("make") || m.equals("<init>") || m.equals("rebase")) continue;
      }
      break;
    }

    // print each level of the stack trace
    out.indent(indent).writeChars(toStr()).writeChar('\n');
    int num = 0;
    for (int i=start; i<elems.length; ++i)
    {
      if (trace(elems[i], out, indent+2)) num++;
      if (num >= maxDepth)
      {
        int more = elems.length - i - start;
        if (more > 0)
          out.indent(indent+2).writeChars(more + " More...\n");
        break;
      }
    }
    out.flush();

    // if this was a rebase, then dump original stack trace
    if (useActual && actual instanceof RebaseException)
      trace(out, opt, indent+2, false);

    // if there is a cause, then recurse
    if (cause != null)
    {
      out.indent(indent).writeChars("Cause:\n");
      cause.trace(out, opt, indent+2, true);
    }

    return this;
  }

  public static boolean trace(StackTraceElement elem, OutStream out, int indent)
  {
    String className  = elem.getClassName();
    String methodName = elem.getMethodName();
    String fileName = elem.getFileName();
    int line = elem.getLineNumber();

    // skip crap like reflection internals
    if (className.startsWith("sun.reflect.")) return false;

    // fan class
    if (className.startsWith("fan.") && !className.startsWith("fan.sys."))
    {
      String podName  = "?";
      String typeName = className;
      String slotName = methodName;

      // map Java full qualified name to pod::type
      int dot = className.indexOf('.', 5);
      if (dot > 0)
      {
        podName  = className.substring(4, dot);
        typeName = className.substring(dot+1);

        // check for closures
        int dollar1 = typeName.indexOf('$');
        int dollar2 = dollar1 < 0 ? -1 : typeName.indexOf('$', dollar1+1);
        if (dollar2 > 0)
        {
          // don't print callX for closures
          if (slotName.startsWith("call")) return false;
          // remap closure class back to original method
          if (slotName.startsWith("doCall"))
          {
            slotName = typeName.substring(dollar1+1, dollar2);
            typeName = typeName.substring(0, dollar1);
          }
        }
      }

      out.indent(indent).writeChars(podName).writeChar(':').writeChar(':')
         .writeChars(typeName).writeChar('.').writeChars(slotName);
    }

    // java class
    else
    {
      out.indent(indent).writeChars(className)
         .writeChar('.').writeChars(methodName).writeChars("");
    }

    // source
    out.writeChar(' ').writeChar('(');
    if (fileName == null) out.writeChars("Unknown");
    else out.writeChars(fileName);
    if (line > 0) out.writeChar(':').writeChars(String.valueOf(line));
    out.writeChar(')').writeChar('\n');
    return true;
  }

  public String traceToStr()
  {
    Buf buf = new MemBuf(1024);
    trace(buf.out());
    return buf.flip().readAllStr();
  }

//////////////////////////////////////////////////////////////////////////
// Interop
//////////////////////////////////////////////////////////////////////////

  public Throwable toJava()
  {
    if (actual != null) return actual;
    return this;
  }

//////////////////////////////////////////////////////////////////////////
// Rebasing
//////////////////////////////////////////////////////////////////////////

  public Err rebase()
  {
    actual = new RebaseException();
    return this;
  }

  public static class RebaseException extends RuntimeException
  {
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  String msg = "";
  Err cause;
  Throwable actual;

}